### PR TITLE
PLANNER-2048 Remove xStreamAnnotatedClass

### PR DESCRIPTION
**Pull request by rsynek**
Motivation: `optaplanner-benchmark` should not be coupled with any data serialization technology, if there is already a general way of reading input datasets, the `SolutionFileIO`.

- [x]  review and update documentation.
- [x]  add a note to the upgrade recipe

Upgrade recipe: https://github.com/kiegroup/optaplanner-website/pull/255

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/cloudbalancing/persistence/CloudBalanceSolutionFileIO.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.cloudbalancing.persistence;
+
+import org.optaplanner.examples.cloudbalancing.domain.CloudBalance;
+import org.optaplanner.persistence.xstream.impl.domain.solution.XStreamSolutionFileIO;
+
+public class CloudBalanceSolutionFileIO extends XStreamSolutionFileIO<CloudBalance> {
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/cloudbalancing/persistence/CloudBalanceSolutionFileIO.java:22]: Are these actually necessary? Maybe there are in the XML, but in the code, I was always able to get away with just using `XStreamSolutionFileIO`.

File: optaplanner-benchmark/src/test/resources/org/optaplanner/benchmark/config/testBenchmarkConfig.xml [hunk 2]
```diff
@@ -44,7 +44,7 @@
       </localSearch>
     </solver>
     <problemBenchmarks>
-      <xStreamAnnotatedClass>org.optaplanner.core.impl.testdata.domain.TestdataSolution</xStreamAnnotatedClass>
+      <solutionFileIOClass>org.optaplanner.benchmark.config.PlannerBenchmarkConfigTest$TestdataSolutionFileIO</solutionFileIOClass>
```
Comment (triceo) [inline optaplanner-benchmark/src/test/resources/org/optaplanner/benchmark/config/testBenchmarkConfig.xml:47]: Can we make it a top-level class? This dollar syntax just makes it more confusing to me.
(Although it's nice to see that it works.)

File: optaplanner-examples/src/main/resources/org/optaplanner/examples/cloudbalancing/optional/benchmark/cloudBalancingBenchmarkConfig.xml [hunk 3]
```diff
@@ -60,95 +60,95 @@
...
-        <!--<exhaustiveSearchType>BRANCH_AND_BOUND</exhaustiveSearchType>-->
-      <!--</exhaustiveSearch>-->
-    <!--</solver>-->
+  <!--<name>BRANCH_AND_BOUND</name>-->
+  <!--<solver>-->
+  <!--<exhaustiveSearch>-->
+  <!--<exhaustiveSearchType>BRANCH_AND_BOUND</exhaustiveSearchType>-->
+  <!--</exhaustiveSearch>-->
+  <!--</solver>-->
   <!--</solverBenchmark>-->
```
Comment (triceo) [inline optaplanner-examples/src/main/resources/org/optaplanner/examples/cloudbalancing/optional/benchmark/cloudBalancingBenchmarkConfig.xml:153]: These changes appear unnecessary.

Comment (rsynek) [reply optaplanner-examples/src/main/resources/org/optaplanner/examples/cloudbalancing/optional/benchmark/cloudBalancingBenchmarkConfig.xml:153]: Sorry, automatic formatting via IDE which escaped my attention. Will remove it.

File: optaplanner-benchmark/src/test/resources/org/optaplanner/benchmark/config/testBenchmarkConfig.xml (hunk 2 above)
Comment (rsynek) [reply optaplanner-benchmark/src/test/resources/org/optaplanner/benchmark/config/testBenchmarkConfig.xml:47]: Sure we can, but the class has nothing to offer outside the test case. I already use several nested classes for the `SolverConfigTest` and would rather not make them standalone top-level classes, as they basically do nothing.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/cloudbalancing/persistence/CloudBalanceSolutionFileIO.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/cloudbalancing/persistence/CloudBalanceSolutionFileIO.java:22]: If you use the `XStreamSolutionFileIO` directly to read a data set, sure. But the benchmark config instantiates the `SolutionFileIO` class itself.

As the constructor requires a user to pass classes that are supposed to be serialized:
```
XStreamSolutionFileIO(Class... xStreamAnnotatedClasses)
```
we would need another config code that reads these classes. And again, we are coupling the benchmark with a subset of possible `SolutionFileIO` implementations.

**Comment by rsynek** [comment]
> Also needs an upgrade recipe entry.

Sure thing, already working on it - see the TODOs in a description of this PR.

**Comment by rsynek** [comment]
jenkins retest this

**Comment by rsynek** [comment]
How do I do it that I hit exactly 80% coverage (the threshold) :) ? Merging.

**Pull request merged**